src-control/Control/RefCount.hs

```diff
@@ -131,3 +165,231 @@ readRefCount RefCounter{countVar} = RefCount <$> readPrimVar countVar
+                                PrimMonad m, MonadMask m)
+                            => obj -> m ()
+  unsharedFinaliser obj =
+    mask_ $ do
+      done <- decrementRefCounter (getRefCounter obj)
+      when done (sharedFinaliser obj)
+
+-- | Class of objects that support 'Ref' including 'dupRef', with the guarantee
+-- that the object's 'sharedFinaliser' will be called exactly once.
+--
+-- Do not define 'unsharedFinaliser' for types in this class. The default
+-- implementation is used for the reference counting.
+--
+class HasSharedFinaliser obj where
+  getRefCounter   :: obj -> RawRefCounter (FinaliserM obj)
+
+  sharedFinaliser :: FinaliserM obj ~ m => obj -> m ()
+
+newRef :: PrimMonad m => obj -> m (Ref obj)
+#ifndef NO_IGNORE_ASSERTS
+newRef obj = return (Ref obj)
+#else
+newRef obj = do
+    released <- unsafeIOToPrim $ newIORef False
+    return (Ref obj released)
+#endif
+
+-- | Convenience helper to make an object containing a reference count.
+newSharedRef :: (HasSharedFinaliser obj, FinaliserM obj ~ m, PrimMonad m)
+             => (RawRefCounter m -> obj) -> m (Ref obj)
+newSharedRef mkObject = do
+    rc  <- newRawRefCounter
+    let !obj = mkObject rc
+    assert (getRefCounter obj == rc) $
+      newRef obj
+
+-- | Release a reference to an object that will no longer be used (via this
+-- reference).
+--
+releaseRef :: (HasFinaliser obj, FinaliserM obj ~ m, PrimMonad m)
+           => Ref obj -> m ()
+#ifndef NO_IGNORE_ASSERTS
+releaseRef (Ref obj) = unsharedFinaliser obj
+#else
+releaseRef (Ref obj released) = do
+    assertNotReleased released
+    unsharedFinaliser obj
+    unsafeIOToPrim $ writeIORef released True
+#endif
+
+-- | Use the object in a 'Ref'. Do not retain the object after the scope of
+-- the body. If you cannot use scoped \"with\" style, use 'unsafeDeRef'.
+--
+#ifndef NO_IGNORE_ASSERTS
+withRef :: Ref obj -> (obj -> m a) -> m a
+withRef (Ref obj) f = f obj
+#else
+withRef :: PrimMonad m => Ref obj -> (obj -> m a) -> m a
+withRef (Ref obj released) f = assertNotReleased released
+                            >> f obj
+#endif
+
+-- | Get the object in a 'Ref'. Be careful with retaining the object for too
+-- long, since the object must not be used after 'releaseRef' is called.
+--
+unsafeDeRef :: Ref obj -> obj
+#ifndef NO_IGNORE_ASSERTS
+unsafeDeRef (Ref obj) = obj
+#else
+unsafeDeRef (Ref obj released) =
+          unsafeDupablePerformIO (assertNotReleased released)
+    `seq` obj
+#endif
+
+-- | Duplicate an existing reference, to produce a new reference.
+--
+dupRef :: (HasSharedFinaliser obj, FinaliserM obj ~ m, PrimMonad m)
+       => Ref obj -> m (Ref obj)
+#ifndef NO_IGNORE_ASSERTS
+dupRef (Ref obj) = do
+    incrementRefCounter (getRefCounter obj)
+    return (Ref obj)
+#else
+dupRef (Ref obj released) = do
+    assertNotReleased released
+    incrementRefCounter (getRefCounter obj)
+    released' <- unsafeIOToPrim $ newIORef False
+    return (Ref obj released')
```

Comment:
This part is just `newRef`, which even already does the CPP, so potentially only `assertNotReleased` would need to be conditional (but only CPP-ing half of the function might not make the code clearer).

Reply:
I've factored it out into a `newRefTracker`.